docs/guides-adv_invokeconcepts.md

```diff
@@ -0,0 +1,63 @@
+---
+id: adv_invokeconcepts
+title: Advanced - Additional doInvoke functions
+---
+
+In this guide, we will discuss features of `doInvoke` that seem a little more complex. We will add an extra transaction to our invoke with `intents` and externalize the signing of our transaction - so we don't have to pass a user's private key to the config.
+
+## Intents - adding extra transactions
+In the previous `doInvoke` guide, we sent 1 GAS alongside our invocation as a fee. As described [here](http://docs.neo.org/en-us/sc/systemfees.html#smart-contract-fees), transactions with a cost under 10 GAS are essentially free. Unfortunately we can't change the gas parameter of our configuration object to 0, since it has to be an integer value greater than 0.  
+
+But there is a way to bypass this. We can add a transaction to ourselves as an intent, with a minimum amount of 0.00000001 GAS. This makes sure we don't have to send 1 GAS to our transaction (without that GAS being needed to consume the transaction), while making sure a transaction is registered and persisted to the blockchain.
+
+The config object that `doInvoke` needs allows for a child object name intents. We set up our intents array as follows:
+```js
+const intents = [
+  {
+    assetId: assets.GAS,
+    value: 0.00000001,
+    scriptHash: Neon.get.scriptHashFromAddress(account.address)
+  }
+];
+
+// Add to config
+config.intents = intents;
+
+Neon.doInvoke(config).then(res => {
+  console.log(res);
+});
+```
+
+## signingFunction
+Right now we're adding a user's private key to the config object, which is sensitive information and should be handled carefully.  
+One way to do so is to externalize the signing of the transaction in a separate function.
+
+Instead of sending a user's private key to the config object, we can send the public key and a function that will sign the transaction.  
+This function, the signingFunction, will receive the transaction and public key as parameters. Now, we can provide logic that retrieves the private key from our user - using the public key to do so - and signs the transaction when we retrieve the key.
+
+> Do note that the signing function has to return a Promise!
+
+```js
+
+function signTx(tx, publicKey) {
+  // Create logic that gets the privateKey based on the publicKey
+  const privateKey = getPrivateKey(publicKey);
+
+  return new Promise(resolve =>
+    resolve(Neon.sign(tx, privateKey))
+  );
+}
+
+const config = {
+  net: "http://localhost:5000",
+  script: Neon.create.script(props),
```

Comment:
Hi Thanks for the doc!

I wanted to know what is the format of props.

I believe it is operation and arguments but could you clarify?

Thans.

Reply:
Hi @melvin0008, you can find the description of `props` right here: http://cityofzion.io/neon-js/docs/en/basic_createscript.html

Reply:
you can dump the `props` directly as the `script` and `doInvoke` will handle it.

Reply:
@snowypowers what would be the best standard mention then? Like you mentioned here https://github.com/CityOfZion/neon-js/pull/191#discussion_r174343021 ?

Or keep it this way and address the various ways of using the script (as neon.create.script vs contractparams object)?

Reply:
```js
const config = {
  net: "",
  script: {
    scriptHash: '5b7074e873973a6ed3708862f219a6fbf4d1c411',
    operation: 'balanceOf',
    args: [Neon.u.reverseHex('cef0c0fdcfe7838eff6ff104f9cdec2922297537')]
  },
  address: account.address,
  publicKey: account.publicKey,
  signingFunction: signTx
  gas: 1
};
```

Reply:
preference is to just put in the script object as thats one less method the user has to know to use `doInvoke`.

Reply:
@jeroenptrs I tried using your code sample and still getting 1 gas deducted every doInvoke call.

Reply:
@vicraft18 that's because this code example has `gas: 1` attached to it. Check the intents example ;)